drivers/usb/usbuhci_new/roothub.c

```diff
@@ -0,0 +1,470 @@
+
+    PortControl.ConnectStatusChange = FALSE;
+    PortControl.PortEnableDisableChange = FALSE;
+
+    if (IsSet)
+    {
+        PortControl.PortEnabledDisabled = TRUE;
+    }
+    else
+    {
+        PortControl.PortEnabledDisabled = FALSE;
+    }
+
+    WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHSetFeaturePortEnable(IN PVOID uhciExtension,
+                           IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    DPRINT("UhciRHSetFeaturePortEnable: ...\n");
+    return UhciRHPortEnable(UhciExtension, Port, TRUE);
+}
+
+MPSTATUS
+NTAPI
+UhciRHSetFeaturePortSuspend(IN PVOID uhciExtension,
+                            IN USHORT Port)
+{
+    DPRINT("UhciRHSetFeaturePortSuspend: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortEnable(IN PVOID uhciExtension,
+                             IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    DPRINT("UhciRHClearFeaturePortEnable: ...\n");
+    return UhciRHPortEnable(UhciExtension, Port, FALSE);
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortPower(IN PVOID uhciExtension,
+                            IN USHORT Port)
+{
+    DPRINT("UhciRHClearFeaturePortPower: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortSuspend(IN PVOID uhciExtension,
+                              IN USHORT Port)
+{
+    DPRINT("UhciRHClearFeaturePortSuspend: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortEnableChange(IN PVOID uhciExtension,
+                                   IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+
+    DPRINT("UhciRHClearFeaturePortEnableChange: ...\n");
+
+    BaseRegister = UhciExtension->BaseRegister;
+    PortControlRegister = (PUSHORT)&BaseRegister->PortControl[Port - 1];
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    PortControl.ConnectStatusChange = FALSE;
+    PortControl.PortEnableDisableChange = TRUE;
+    WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortConnectChange(IN PVOID uhciExtension,
+                                    IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+
+    DPRINT("UhciRHClearFeaturePortConnectChange: Port - %04X\n", Port);
+
+    BaseRegister = UhciExtension->BaseRegister;
+    PortControlRegister = (PUSHORT)&BaseRegister->PortControl[Port - 1];
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    if (PortControl.ConnectStatusChange == TRUE)
```

Comment:
Why do you check whether the bit is set here, but not in UhciRHClearFeaturePortEnableChange?

Reply:
Why does it need to be checked in UhciRHClearFeaturePortEnableChange?

Reply:
Why does it need to be checked here? My point is that these two functions basically do the same thing, just with different bits. So it seems odd that one has the check and one does not.

Reply:
It makes no sense to do a check in UhciRHClearFeaturePortEnableChange(). For UhciRHClearFeaturePortEnableChange() such a case is possible when we call USBH_SyncClearPortStatus() in USBH_StartHubFdoDevice().